### PR TITLE
Issue 1044 account deletion

### LINKED ISSUE #468: Account deletion
There is currently no option to delete a member's account.  As a member, I should be able to do this via my settings page.

We need to make sure that all related things are deleted (the member's plantings, gardens, photos, etc) but also make sure that:
- if the member is a crop wrangler, don't delete crops they've added -- just reassign them to cropbot or something
- if the member is or was an admin, ditto.
- we need to figure out what to do with deleted posts -- maybe replace them with "DELETED POST" so you can see that something was there, but not the details of what it was?  This is what Dreamwidth does.  If you don't do this, then any replies look very odd indeed.

### PR DESCRIPTION
**Pull request by Br3nda**
Uses acts_as_paranoid to mark members as deleted.

~~If a query does `Planting.all` etc, it will get all plantings, but some will have no owner.
To only see plantings with owners, do `Plantings.joins(:owner).all`~~
The default scopes only allows see gardens+plantings+Harvests+seeds with a not-deleted owner. 

includes code from #1340 because it needs the interesting seeds scope fixed.
this includes some code from #743 
fixes #468

- [x] Feature spec of pressing the delete account button
- [x] Check password matches before allowing deleting
- [x] Soft delete user's data too. (Gardens, Plantings, Harvests, Seeds).

### CODE REVIEW

File: app/controllers/members_controller.rb [hunk 1]
```diff
@@ -54,6 +54,11 @@ def show
     end
   end
 
+  def destroy
```
Comment (Br3nda) [inline app/controllers/members_controller.rb]: TODO!! Needs to check the password supplied is correct before letting them destroy their account.

**Comment by cesy** [comment]
Thank you so much for picking this up.

File: db/migrate/20170413221549_counter_caches.rb [hunk 2]
```diff
@@ -4,7 +4,7 @@ def change
     add_column :members, :harvests_count, :integer
     add_column :members, :seeds_count, :integer
 
-    Member.find_each do |member|
+    Member.unscoped.find_each do |member|
```
Comment (Br3nda) [inline db/migrate/20170413221549_counter_caches.rb]: I had to modify this old migration. The models now want the deleted_at column to exist, and it doesn't yet at the point in the migration.

**Comment by Br3nda** [comment]
I need some decisions in whether it's okay to expose a deleted users gardens and plantings and photos and seeds. This data will appear to have no owner. I'm not sure this is what a user expects when they delete their account. Should I soft delete their other data too?

**Comment by pozorvlak** [comment]
Yes, I think soft-deleting their other data is the Right Thing.

**Comment by Br3nda** [comment]
Interestingly, the DELETE member action is already there -- no controller changes needed. I couldn't immediately work out how to override that - probably need to read up on devise more - hence this PR is still a work in progress.

**Comment by Br3nda** [comment]
Added specs from @cesy 's  61f93df6ff5ef9bf74606a90a262f6db41092f09
Now checks password was provided before allowing delete
Changed order.member's belongs_to to include deleted
Included deleted members on the order search

Ready for review.

File: app/controllers/plantings_controller.rb [hunk 3]
```diff
@@ -12,7 +12,7 @@ def index
     @crop = Crop.find_by(slug: params[:crop]) if params[:crop]
     @show_all = params[:all] == '1'
 
-    @plantings = plantings.paginate(page: params[:page])
+    @plantings = plantings
```
Comment (Br3nda) [inline app/controllers/plantings_controller.rb]: oh, i saw this was already paginated, so i removed the duplicate while here.

**Comment by Br3nda** [comment]
I'd really love a review here - so we can get the release out...
anyone?

**Comment by cesy** [comment]
I'll have a look when I can, but don't have the brain for a full review at the moment, so may need help from others.

File: spec/features/members/deletion_spec.rb [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+      login_as(member)
+      FactoryGirl.create(:comment, author: member, post: othermemberpost)
+      FactoryGirl.create(:comment, author: other_member, post: memberpost, body: "Fun comment-y thing")
+      # deletion breaks if no wranglers exist
+      FactoryGirl.create(:cropbot)
+      # deletion breaks if ex_member doesn't exist
+      FactoryGirl.create(:member, login_name: "ex_member")
+    end
+
+    # scenario "has option to delete on member profile page" do
```
Comment (cesy) [inline spec/features/members/deletion_spec.rb]: Does this one still need fixing?

File: spec/features/members/deletion_spec.rb [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+      scenario "replaces posts with deletion note" do
+        visit post_path(memberpost)
+        expect(page.status_code).to eq(404)
+      end
+
+      scenario "replaces comments on others' posts with deletion note, leaving post intact" do
+        visit post_path(othermemberpost)
+        expect(page).not_to have_content member.login_name.to_s
+        # TODO
```
Comment (cesy) [inline spec/features/members/deletion_spec.rb]: Still outstanding?

File: spec/features/members/deletion_spec.rb [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        # expect(page).to have_content other_member.login_name.to_s
+        # expect(page).to have_content "This comment was removed as the author deleted their account."
+      end
+
+      scenario "leaves a record of orders and payments intact" do
```
Comment (cesy) [inline spec/features/members/deletion_spec.rb]: Thank you for covering this.

**Comment by cesy** [comment]
Thank you so much for pushing this forward.

**Comment by Br3nda** [comment]
Pending specs added -- and more model specs.
added a before_destroy hook to unsubscribe from newsletter

**Comment by cesy** [comment]
I'd like someone to run a manual test on staging before we put this onto production.

**Pull request merged**